File: packages/nativephp/native-ui/resources/ios/NativeUITheme.swift
import SwiftUI
import UIKit

// MARK: - Theme Tokens

/// One coherent set of theme tokens (light OR dark). Exposed to SwiftUI
/// renderers via the `\.nativeUITheme` environment value.
///
/// The PHP side (`Nativephp\NativeUi\Theme::merge([...])`) is the source of
/// truth. Tokens arrive over the bridge via `NativeUI.Theme.Set`.
struct NativeUITokens: Equatable {
    // Colors
    let primary: Color
    let onPrimary: Color
    let secondary: Color
    let onSecondary: Color
    let surface: Color
    let onSurface: Color
    let background: Color
    let onBackground: Color
    let surfaceVariant: Color
    let onSurfaceVariant: Color
    let outline: Color
    let destructive: Color
    let onDestructive: Color
    let accent: Color
    let onAccent: Color

    // Radii (points)
    let radiusSm: CGFloat
    let radiusMd: CGFloat
    let radiusLg: CGFloat
    let radiusFull: CGFloat

    // Typography
    let fontSm: CGFloat
    let fontMd: CGFloat
    let fontLg: CGFloat
    let fontXl: CGFloat
    let fontFamily: String

    /// Fallback matches `config/native-ui.php` defaults. Used before PHP has
    /// pushed an initial theme — keeps components renderable during bootstrap.
    static let fallback = NativeUITokens(
        primary:          Color(hex: "#0F766E"),
        onPrimary:        Color(hex: "#FFFFFF"),
        secondary:        Color(hex: "#475569"),
        onSecondary:      Color(hex: "#FFFFFF"),
        surface:          Color(hex: "#FFFFFF"),
        onSurface:        Color(hex: "#0F172A"),
        background:       Color(hex: "#F8FAFC"),
        onBackground:     Color(hex: "#0F172A"),
        surfaceVariant:   Color(hex: "#F1F5F9"),
        onSurfaceVariant: Color(hex: "#475569"),
        outline:          Color(hex: "#CBD5E1"),
        destructive:      Color(hex: "#B91C1C"),
        onDestructive:    Color(hex: "#FFFFFF"),
        accent:           Color(hex: "#C2410C"),
        onAccent:         Color(hex: "#FFFFFF"),
        radiusSm: 4, radiusMd: 8, radiusLg: 16, radiusFull: 9999,
        fontSm: 14, fontMd: 16, fontLg: 20, fontXl: 24,
        fontFamily: "System"
    )
}

// MARK: - Theme Store

/// Singleton holding the current effective theme tokens for both color schemes.
/// PHP pushes updates via `NativeUI.Theme.Set`; consumers pick the right scheme
/// via `resolve(for: colorScheme)`.
final class NativeUITheme: ObservableObject {
    static let shared = NativeUITheme()

    @Published private(set) var light: NativeUITokens = .fallback
    @Published private(set) var dark: NativeUITokens = .fallback

    private init() {}

    /// Pick the active token set based on the current system color scheme.
    func resolve(for scheme: ColorScheme) -> NativeUITokens {
        scheme == .dark ? dark : light
    }

    /// Called by the bridge on every `Theme::merge()` from PHP. Parameters are
    /// the full effective theme (light + dark + radii + fonts), post-merge
    /// and post-dark-auto-derivation.
    func apply(_ parameters: [String: Any]) {
        let lightMap = parameters["light"] as? [String: Any] ?? [:]
        let darkMap  = parameters["dark"]  as? [String: Any] ?? [:]

        let radiusSm   = cgf(parameters["radius-sm"],   fallback: light.radiusSm)
        let radiusMd   = cgf(parameters["radius-md"],   fallback: light.radiusMd)
        let radiusLg   = cgf(parameters["radius-lg"],   fallback: light.radiusLg)
        let radiusFull = cgf(parameters["radius-full"], fallback: light.radiusFull)

        let fontSm     = cgf(parameters["font-sm"],     fallback: light.fontSm)
        let fontMd     = cgf(parameters["font-md"],     fallback: light.fontMd)
        let fontLg     = cgf(parameters["font-lg"],     fallback: light.fontLg)
        let fontXl     = cgf(parameters["font-xl"],     fallback: light.fontXl)
        let fontFamily = (parameters["font-family"] as? String) ?? light.fontFamily

        func tokens(from map: [String: Any], fallbackTo: NativeUITokens) -> NativeUITokens {
            NativeUITokens(
                primary:          hex(map["primary"],            fallback: fallbackTo.primary),
                onPrimary:        hex(map["on-primary"],         fallback: fallbackTo.onPrimary),
                secondary:        hex(map["secondary"],          fallback: fallbackTo.secondary),
                onSecondary:      hex(map["on-secondary"],       fallback: fallbackTo.onSecondary),
                surface:          hex(map["surface"],            fallback: fallbackTo.surface),
                onSurface:        hex(map["on-surface"],         fallback: fallbackTo.onSurface),
                background:       hex(map["background"],         fallback: fallbackTo.background),
                onBackground:     hex(map["on-background"],      fallback: fallbackTo.onBackground),
                surfaceVariant:   hex(map["surface-variant"],    fallback: fallbackTo.surfaceVariant),
                onSurfaceVariant: hex(map["on-surface-variant"], fallback: fallbackTo.onSurfaceVariant),
                outline:          hex(map["outline"],            fallback: fallbackTo.outline),
                destructive:      hex(map["destructive"],        fallback: fallbackTo.destructive),
                onDestructive:    hex(map["on-destructive"],     fallback: fallbackTo.onDestructive),
                accent:           hex(map["accent"],             fallback: fallbackTo.accent),
                onAccent:         hex(map["on-accent"],          fallback: fallbackTo.onAccent),
                radiusSm: radiusSm, radiusMd: radiusMd, radiusLg: radiusLg, radiusFull: radiusFull,
                fontSm: fontSm, fontMd: fontMd, fontLg: fontLg, fontXl: fontXl,
                fontFamily: fontFamily
            )
        }

        let newLight = tokens(from: lightMap, fallbackTo: .fallback)
        let newDark  = tokens(from: darkMap,  fallbackTo: newLight)

        // `@Published` fires on every assignment — even when the new value
        // equals the old one. That would cause every observer (every Screen,
        // Button, TextInput…) to re-render on each bridge push, which PHP
        // sends on every service-provider boot (i.e. every request). Gate the
        // assignment on actual change to keep the observable stable.
        if newLight != self.light { self.light = newLight }
        if newDark  != self.dark  { self.dark  = newDark }

        applyChromeFont(fontFamily)
    }

    /// Font the SYSTEM-drawn navigation chrome. `navigationTitle` renders
    /// through UIKit's `UINavigationBar` and SwiftUI exposes no font API for
    /// it, so route the app-wide default font through the UIKit appearance
    /// proxy. Core doesn't configure `UINavigationBar` appearances itself, so
    /// this composes safely; bars pick it up on creation (theme lands at boot,
    /// before the first screen). Tab-bar labels are excluded — core's tabs
    /// renderer manages its own `UITabBarAppearance`.
    private func applyChromeFont(_ family: String) {
        if family.isEmpty || family == "System" {
            UINavigationBar.appearance().titleTextAttributes = nil
            UINavigationBar.appearance().largeTitleTextAttributes = nil
            return
        }

        guard let psName = NativeUIFontResolver.postScriptName(for: family) else { return }

        if let title = UIFont(name: psName, size: 17) {
            UINavigationBar.appearance().titleTextAttributes = [.font: title]
        }
        if let large = UIFont(name: psName, size: 34) {
            UINavigationBar.appearance().largeTitleTextAttributes = [.font: large]
        }
    }
}

// MARK: - Environment Propagation

private struct NativeUIThemeKey: EnvironmentKey {
    static let defaultValue: NativeUITokens = .fallback
}

extension EnvironmentValues {
    /// Active theme tokens for the current color scheme. Renderers read this
    /// instead of going through `NativeUITheme.shared` directly so SwiftUI can
    /// track dependencies and update components when tokens change.
    var nativeUITheme: NativeUITokens {
        get { self[NativeUIThemeKey.self] }
        set { self[NativeUIThemeKey.self] = newValue }
    }
}

// MARK: - Parsing helpers

private func hex(_ any: Any?, fallback: Color) -> Color {
    guard let s = any as? String, s.hasPrefix("#") else { return fallback }
    return Color(hex: s)
}

private func cgf(_ any: Any?, fallback: CGFloat) -> CGFloat {
    if let n = any as? CGFloat { return n }
    if let n = any as? Double { return CGFloat(n) }
    if let n = any as? Int { return CGFloat(n) }
    if let n = any as? NSNumber { return CGFloat(truncating: n) }
    return fallback
}

// MARK: - Accessibility helpers

/// Renders a fixed design-point font size through Dynamic Type so text tracks
/// the user's preferred content size. `@ScaledMetric` scales the base size
/// relative to `.body`, matching how UIFontMetrics scales custom fonts.
struct NUIScaledFontModifier: ViewModifier {
    @ScaledMetric private var size: CGFloat
    private let weight: Font.Weight
    private let design: Font.Design
    private let fontName: String?

    init(size: CGFloat, weight: Font.Weight, design: Font.Design, fontName: String?) {
        _size = ScaledMetric(wrappedValue: size, relativeTo: .body)
        self.weight = weight
        self.design = design
        self.fontName = fontName
    }

    @ObservedObject private var themeStore = NativeUITheme.shared
    @Environment(\.colorScheme) private var colorScheme

    func body(content: Content) -> some View {
        // A resolvable custom font wins; the weight still applies (SwiftUI
        // selects/synthesizes it within the family). Unknown names — or none —
        // fall back to the system font unchanged. `size` is already Dynamic-
        // Type-scaled by `@ScaledMetric`, so the custom font uses it as-is.
        if let name = effectiveFontName, let custom = NativeUIFontResolver.font(name, size: size) {
            content.font(custom.weight(weight))
        } else {
            content.font(.system(size: size, weight: weight, design: design))
        }
    }

    /// Explicit per-element font first; otherwise the app-wide default from
    /// the theme's `font-family` token — but only for the default design, so
    /// explicit `font-serif` / `font-mono` classes still win over the default.
    private var effectiveFontName: String? {
        if let fontName { return fontName }
        guard design == .default else { return nil }

        let family = themeStore.resolve(for: colorScheme).fontFamily
        return (family.isEmpty || family == "System") ? nil : family
    }
}

extension View {
    /// Dynamic-Type-aware replacement for `.font(.system(size:weight:design:))`.
    /// Use this for all text (and glyph icons rendered as text) so type scales
    /// with the user's accessibility settings. Pass `fontName` (a bundled font
    /// token) to render in a custom font, falling back to the system font when
    /// it can't be resolved.
    func nuiScaledFont(size: CGFloat, weight: Font.Weight = .regular, design: Font.Design = .default, fontName: String? = nil) -> some View {
        modifier(NUIScaledFontModifier(size: size, weight: weight, design: design, fontName: fontName))
    }

    /// Ensures a minimum 44×44pt hit target (Apple HIG) without changing the
    /// visual size of the content. Apply to small tappable controls.
    func nuiMinTapTarget() -> some View {
        frame(minWidth: 44, minHeight: 44)
            .contentShape(Rectangle())
    }
}

extension Color {
    /// Construct from `#RRGGBB` or `#AARRGGBB` hex string. Falls back to black
    /// on malformed input.
    init(hex: String) {
        var s = hex
        if s.hasPrefix("#") { s.removeFirst() }
        s = s.uppercased()

        var argb: UInt64 = 0xFF000000
        if s.count == 6, let rgb = UInt64(s, radix: 16) {
            argb = 0xFF000000 | rgb
        } else if s.count == 8, let full = UInt64(s, radix: 16) {
            argb = full
        }

        let a = Double((argb >> 24) & 0xFF) / 255.0
        let r = Double((argb >> 16) & 0xFF) / 255.0
        let g = Double((argb >>  8) & 0xFF) / 255.0
        let b = Double((argb >>  0) & 0xFF) / 255.0
        self = Color(red: r, green: g, blue: b, opacity: a)
    }
}

File: packages/nativephp/native-ui/resources/android/NativeUITheme.kt
package com.nativephp.plugins.native_ui

import androidx.compose.material3.ColorScheme
import androidx.compose.material3.darkColorScheme
import androidx.compose.material3.lightColorScheme
import androidx.compose.runtime.compositionLocalOf
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.TextUnit
import androidx.compose.ui.unit.sp
import org.json.JSONObject

/**
 * One coherent set of theme tokens (light OR dark). Renderers read the
 * active tokens via [LocalNativeUITheme].
 *
 * The PHP side (`Nativephp\NativeUi\Theme::merge([...])`) is the source of
 * truth; tokens arrive over the bridge via `NativeUI.Theme.Set`.
 */
data class NativeUITokens(
    // Colors
    val primary:          Color,
    val onPrimary:        Color,
    val secondary:        Color,
    val onSecondary:      Color,
    val surface:          Color,
    val onSurface:        Color,
    val background:       Color,
    val onBackground:     Color,
    val surfaceVariant:   Color,
    val onSurfaceVariant: Color,
    val outline:          Color,
    val destructive:      Color,
    val onDestructive:    Color,
    val accent:           Color,
    val onAccent:         Color,

    // Radii
    val radiusSm:   Dp,
    val radiusMd:   Dp,
    val radiusLg:   Dp,
    val radiusFull: Dp,

    // Typography
    val fontSm:     TextUnit,
    val fontMd:     TextUnit,
    val fontLg:     TextUnit,
    val fontXl:     TextUnit,
    val fontFamily: String,
) {
    companion object {
        /**
         * Fallback matches `config/native-ui.php` defaults. Used before PHP
         * has pushed an initial theme — keeps components renderable during
         * bootstrap.
         */
        val fallback = NativeUITokens(
            primary          = parseHex("#0F766E"),
            onPrimary        = parseHex("#FFFFFF"),
            secondary        = parseHex("#64748B"),
            onSecondary      = parseHex("#FFFFFF"),
            surface          = parseHex("#FFFFFF"),
            onSurface        = parseHex("#0F172A"),
            background       = parseHex("#F8FAFC"),
            onBackground     = parseHex("#0F172A"),
            surfaceVariant   = parseHex("#F1F5F9"),
            onSurfaceVariant = parseHex("#475569"),
            outline          = parseHex("#CBD5E1"),
            destructive      = parseHex("#DC2626"),
            onDestructive    = parseHex("#FFFFFF"),
            accent           = parseHex("#FB923C"),
            onAccent         = parseHex("#FFFFFF"),
            radiusSm = 4.dp, radiusMd = 8.dp, radiusLg = 16.dp, radiusFull = 9999.dp,
            fontSm = 14.sp, fontMd = 16.sp, fontLg = 20.sp, fontXl = 24.sp,
            fontFamily = "System",
        )
    }
}

/**
 * Process-wide theme store. Observable via `mutableStateOf` so Compose
 * re-composes when the PHP side pushes an update.
 */
object NativeUITheme {
    var light by mutableStateOf(NativeUITokens.fallback)
        private set
    var dark by mutableStateOf(NativeUITokens.fallback)
        private set

    /**
     * Apply an effective token set from PHP. The [parameters] map is the
     * decoded JSON from `Theme::all()` — light + dark blocks plus radii and
     * typography scalars.
     */
    fun apply(parameters: Map<String, Any?>) {
        // Android's bridge layer leaves nested JSON objects as org.json.JSONObject
        // rather than converting them to Map<String, Any> (unlike iOS's Foundation
        // JSONSerialization). Use asMap() to accept either representation.
        val lightMap = asMap(parameters["light"])
        val darkMap  = asMap(parameters["dark"])

        val radiusSm   = numDp(parameters["radius-sm"],   NativeUITokens.fallback.radiusSm)
        val radiusMd   = numDp(parameters["radius-md"],   NativeUITokens.fallback.radiusMd)
        val radiusLg   = numDp(parameters["radius-lg"],   NativeUITokens.fallback.radiusLg)
        val radiusFull = numDp(parameters["radius-full"], NativeUITokens.fallback.radiusFull)

        val fontSm     = numSp(parameters["font-sm"],     NativeUITokens.fallback.fontSm)
        val fontMd     = numSp(parameters["font-md"],     NativeUITokens.fallback.fontMd)
        val fontLg     = numSp(parameters["font-lg"],     NativeUITokens.fallback.fontLg)
        val fontXl     = numSp(parameters["font-xl"],     NativeUITokens.fallback.fontXl)
        val fontFamily = (parameters["font-family"] as? String) ?: NativeUITokens.fallback.fontFamily

        val lightTokens = tokensFrom(
            lightMap, NativeUITokens.fallback,
            radiusSm, radiusMd, radiusLg, radiusFull,
            fontSm, fontMd, fontLg, fontXl, fontFamily
        )
        val darkTokens = tokensFrom(
            darkMap, lightTokens,
            radiusSm, radiusMd, radiusLg, radiusFull,
            fontSm, fontMd, fontLg, fontXl, fontFamily
        )

        // `mutableStateOf` notifies observers on every assignment — even when
        // the new value equals the old one. PHP pushes the theme on every
        // service-provider boot (each request) so unguarded assignment would
        // force every observer to recompose on every Livewire event. Skip
        // the write when tokens haven't actually changed.
        if (lightTokens != light) light = lightTokens
        if (darkTokens  != dark)  dark  = darkTokens
    }

    private fun tokensFrom(
        map: Map<String, Any?>,
        fb: NativeUITokens,
        radiusSm: Dp, radiusMd: Dp, radiusLg: Dp, radiusFull: Dp,
        fontSm: TextUnit, fontMd: TextUnit, fontLg: TextUnit, fontXl: TextUnit,
        fontFamily: String,
    ): NativeUITokens = NativeUITokens(
        primary          = color(map["primary"],            fb.primary),
        onPrimary        = color(map["on-primary"],         fb.onPrimary),
        secondary        = color(map["secondary"],          fb.secondary),
        onSecondary      = color(map["on-secondary"],       fb.onSecondary),
        surface          = color(map["surface"],            fb.surface),
        onSurface        = color(map["on-surface"],         fb.onSurface),
        background       = color(map["background"],         fb.background),
        onBackground     = color(map["on-background"],      fb.onBackground),
        surfaceVariant   = color(map["surface-variant"],    fb.surfaceVariant),
        onSurfaceVariant = color(map["on-surface-variant"], fb.onSurfaceVariant),
        outline          = color(map["outline"],            fb.outline),
        destructive      = color(map["destructive"],        fb.destructive),
        onDestructive    = color(map["on-destructive"],     fb.onDestructive),
        accent           = color(map["accent"],             fb.accent),
        onAccent         = color(map["on-accent"],          fb.onAccent),
        radiusSm = radiusSm, radiusMd = radiusMd, radiusLg = radiusLg, radiusFull = radiusFull,
        fontSm = fontSm, fontMd = fontMd, fontLg = fontLg, fontXl = fontXl,
        fontFamily = fontFamily,
    )
}

/**
 * Composition local exposing the *active* token set. Renderers read this
 * instead of touching [NativeUITheme] directly so Compose can track
 * dependencies and recompose when tokens change.
 */
val LocalNativeUITheme = compositionLocalOf { NativeUITokens.fallback }

/**
 * Derive a Material 3 [ColorScheme] from these plugin tokens.
 *
 * Native chrome on Android (TopAppBar, Scaffold, BottomNav, SideDrawer,
 * any vanilla M3 control) reads from `MaterialTheme.colorScheme.*`, while
 * per-component plugin renderers read from [LocalNativeUITheme]. Without
 * a bridge between the two, chrome stays on M3's baseline (lavender /
 * deep-purple) even after PHP pushes a custom theme via `Theme::merge`.
 *
 * Mapping notes:
 *   - `accent` → M3's `tertiary` slot (closest semantic match).
 *   - `destructive` → M3's `error` slot.
 *   - Container/inverse/scrim slots have no plugin equivalent; we leave
 *     them at M3's baseline (via `[lightColorScheme] / [darkColorScheme]`)
 *     so newly-introduced M3 widgets that depend on them keep working.
 *   - `surfaceTint = primary` so elevation tinting follows the brand.
 *
 * Call from a `@Composable` context that already knows the active mode
 * (system or PHP-overridden). The token instance itself doesn't carry a
 * light/dark flag — it's just one coherent palette.
 */
fun NativeUITokens.toMaterialColorScheme(isDark: Boolean = false): ColorScheme {
    val base = if (isDark) darkColorScheme() else lightColorScheme()
    return base.copy(
        primary          = primary,
        onPrimary        = onPrimary,
        secondary        = secondary,
        onSecondary      = onSecondary,
        tertiary         = accent,
        onTertiary       = onAccent,
        background       = background,
        onBackground     = onBackground,
        surface          = surface,
        onSurface        = onSurface,
        surfaceVariant   = surfaceVariant,
        onSurfaceVariant = onSurfaceVariant,
        outline          = outline,
        error            = destructive,
        onError          = onDestructive,
        surfaceTint      = primary,
        // M3's "surface container" tonal family. These back DropdownMenu,
        // NavigationBar/TabRow, sheets, and the watcher's loading indicator.
        // Left unmapped they keep lightColorScheme()'s baseline (a pinkish
        // neutral), which read as off-theme. Fold them onto the configured
        // surface tokens: the lower tiers track `surface`, the elevated tiers
        // track `surfaceVariant`, so anchored chrome stays on-palette.
        surfaceContainerLowest  = surface,
        surfaceContainerLow     = surface,
        surfaceContainer        = surface,
        surfaceContainerHigh    = surfaceVariant,
        surfaceContainerHighest = surfaceVariant,
        surfaceBright           = surface,
        surfaceDim              = surfaceVariant,
        inverseSurface          = onSurface,
        inverseOnSurface        = surface,
        inversePrimary          = primary,
    )
}

// ─── Parsing helpers ─────────────────────────────────────────────────────────

/// Accept either a real Map (iOS-style bridge payload) or a [JSONObject]
/// (Android bridge default for nested objects) and return a uniform
/// `Map<String, Any?>` for downstream consumption.
private fun asMap(any: Any?): Map<String, Any?> = when (any) {
    is Map<*, *> -> {
        @Suppress("UNCHECKED_CAST")
        (any as? Map<String, Any?>) ?: emptyMap()
    }
    is JSONObject -> {
        val out = mutableMapOf<String, Any?>()
        val keys = any.keys()
        while (keys.hasNext()) {
            val key = keys.next()
            out[key] = if (any.isNull(key)) null else any.opt(key)
        }
        out
    }
    else -> emptyMap()
}

private fun color(any: Any?, fallback: Color): Color =
    (any as? String)?.takeIf { it.startsWith("#") }?.let(::parseHex) ?: fallback

private fun numDp(any: Any?, fallback: Dp): Dp = when (any) {
    is Int    -> any.dp
    is Long   -> any.toInt().dp
    is Double -> any.toFloat().dp
    is Float  -> any.dp
    is Number -> any.toFloat().dp
    else      -> fallback
}

private fun numSp(any: Any?, fallback: TextUnit): TextUnit = when (any) {
    is Int    -> any.sp
    is Long   -> any.toInt().sp
    is Double -> any.toFloat().sp
    is Float  -> any.sp
    is Number -> any.toFloat().sp
    else      -> fallback
}

/** Parse `#RRGGBB` or `#AARRGGBB` into a Compose [Color]. */
private fun parseHex(hex: String): Color {
    val clean = hex.removePrefix("#")
    val long = try { clean.toLong(16) } catch (_: NumberFormatException) { return Color.Black }
    val argb: Long = when (clean.length) {
        6 -> 0xFF000000L or long
        8 -> long
        else -> return Color.Black
    }
    return Color(
        alpha = ((argb shr 24) and 0xFF) / 255f,
        red   = ((argb shr 16) and 0xFF) / 255f,
        green = ((argb shr 8)  and 0xFF) / 255f,
        blue  = ( argb         and 0xFF) / 255f,
    )
}